sdk/webpubsub/azure-webpubsub-client/azure/webpubsub/client/_client.py

```diff
@@ -0,0 +1,760 @@
+        else:
+            raise TypeError("type of credential must be str or WebPubSubClientCredential")
+
+        self._options = options or WebPubSubClientOptions()
+        self._build_default_options()
+        self._message_retry_policy = RetryPolicy(
+            retry_options=self._options.message_retry_options or WebPubSubRetryOptions()
+        )
+        self._reconnect_retry_policy = RetryPolicy(
+            WebPubSubRetryOptions(
+                max_retries=sys.maxsize, retry_delay_in_ms=1000, max_retry_delay_in_ms=30000, mode="Fixed"
+            )
+        )
+        self._protocol = self._options.protocol or WebPubSubJsonReliableProtocol()
+        self._group_map: Dict[str, WebPubSubGroup] = {}
+        self._ack_map: Dict[int, SendMessageErrorOptions] = {}
+        self._sequence_id = SequenceId()
+        self._state = WebPubSubClientState.STOPPED
+        self._ack_id = 0
+        self._url = None
+        self._ws: Optional[websocket.WebSocketApp] = None
+        self._handler: Dict[str, List[Callable]] = {
+            CallBackType.CONNECTED: [],
+            CallBackType.DISCONNECTED: [],
+            CallBackType.REJOIN_GROUP_FAILED: [],
+            CallBackType.GROUP_MESSAGE: [],
+            CallBackType.SERVER_MESSAGE: [],
+            CallBackType.STOPPED: [],
+        }
+        self._last_disconnected_message: Optional[DisconnectedMessage] = None
+        self._connection_id: Optional[str] = None
+        self._is_initial_connected = False
+        self._is_stopping = False
+        self._last_close_event: Optional[CloseEvent] = None
+        self._reconnection_token = None
+        self._cv: threading.Condition = threading.Condition()
+        self._thread_seq_ack: Optional[threading.Thread] = None
+        self._thread: Optional[threading.Thread] = None
+
+    def _next_ack_id(self) -> int:
+        self._ack_id = self._ack_id + 1
+        return self._ack_id
+
+    def _send_message(self, message: WebPubSubMessage):
+        pay_load = self._protocol.write_message(message)
+        if not self._ws or not self._ws.sock:
+            raise Exception("The connection is not connected.")
+
+        self._ws.send(pay_load)
+
+    def _send_message_with_ack_id(
+        self,
+        message_provider: Callable[[int], WebPubSubMessage],
+        ack_id: Optional[int] = None,
+    ):
+        if ack_id is None:
+            ack_id = self._next_ack_id()
+
+        message = message_provider(ack_id)
+        if ack_id not in self._ack_map:
+            self._ack_map[ack_id] = SendMessageErrorOptions(
+                error_detail=AckMessageError(name="", message="timeout to receive ack message")
+            )
+        try:
+            self._send_message(message)
+        except Exception as e:
+            self._ack_map.pop(ack_id)
+            raise e
+
+        # wait for ack from service
+        with self._ack_map[ack_id].cv:
+            self._ack_map[ack_id].cv.wait(60.0)
```

Comment:
I'm wondering whether we need to make magic numbers like 60 here to be configurable for the end user.
One solution is just define them as client variable and be configurable through a `**kwargs` in client.__init__(...)

Reply:
Make sense. I think client variable solution is better.